index.bs

```diff
@@ -368,6 +364,26 @@ The <dfn method for="XRSession">end()</dfn> method provides a way to manually sh
 
 </div>
 
+Each {{XRSession}} has an <dfn>active render state</dfn> which is a new {{XRRenderState}}, a <dfn>pending render state</dfn> which is a new {{XRRenderState}}, and a <dfn>list of pending render state promises</dfn>, which is initially empty.
+
+The <dfn attribute for="XRSession">renderState</dfn> attribute returns the {{XRSession}}'s [=active render state=].
+
+<div class="algorithm" data-algorithm="update-render-state">
+
+When the <dfn method for="XRSession">updateRenderState(|newState|)</dfn> method is invoked, the user agent MUST return [=a new Promise=] |promise| and run the following steps [=in parallel=]:
+
+  1. Let |session| be the target {{XRSession}}.
+  1. Let |pendingState| be |session|'s [=pending render state=].
+  1. If |newState|'s {{XRRenderStateInit/depthNear}} value is set, set |pendingState|'s {{XRRenderState/depthNear}} to |newState|'s {{XRRenderStateInit/depthNear}}.
+  1. If |newState|'s {{XRRenderStateInit/depthFar}} value is set, set |pendingState|'s {{XRRenderState/depthFar}} to |newState|'s {{XRRenderStateInit/depthFar}}.
+  1. If |newState|'s {{XRRenderStateInit/baseLayer}} is set, set |pendingState|'s {{XRRenderState/baseLayer}} to |newState|'s {{XRRenderStateInit/baseLayer}}.
+  1. If |session|'s [=processing frame=] flag is <code>false</code> and |session|'s [=list of animation frame callbacks=] is empty, copy |session|'s [=pending render state=] to |session|'s [=active render state=] and [=/Resolve=] |promise|.
```

Comment:
I'm inclined to have the spec always put the promise in the queue and defer switching the pending render state to be the active render state to the algorithm for invoking the next batch of frame callbacks. wdyt?

Reply:
Sorry, I realized that I had some uncommitted changes with my last code push that change this logic up a bit. You'll want to take a quick look again.

As for always deferring the state change till the next rAF callback, I didn't feel like that was appropriate because you may have a session that you're not actively pumping the render loop for (maybe it's suspended?) that you don't want pending state changes to stack up on forever. The latest algorithm changes I've made state that the changes always immediately apply unless you're in the middle of a batch of frame callbacks, at which point the changes are deferred until the frame callbacks have finished.

Reply:
> that you don't want pending state changes to stack up on forever

You kinda need this anyway, an implementation will need a separate thread adding rAF events to the event loop, and when the session is suspended (e.g. the layer is null) it probably doesn't want to keep triggering rAF and needs some wakeup signaling mechanism.